### PR TITLE
feat: parse server

### PR DESCRIPTION
**Pull request by sogunshola**
implement logic for streams parse server plugin

---
name: 'Pull request'
about: A new pull request
---

### Checklist

- [X] I am not disclosing a [vulnerability](https://github.com/MoralisWeb3/Moralis-JS-SDK/blob/main/SECURITY.md).
- [X] My code is conform the [code style](https://github.com/MoralisWeb3/Moralis-JS-SDK/blob/main/CODE_STYLE.md)
- [X] I have made corresponding changes to the documentation
- [X] I have updated Typescript definitions when needed

### Issue Description

Related issue: #`FILL_THIS_OUT`

### CODE REVIEW

**Review by b4rtaz — CHANGES_REQUESTED** [review]
The logic from the `@moralisweb3/common-streams-utils` package should be moved to the `@moralisweb3/streams` package as Erno recommended.

```
@moralisweb3/streams
|_ /mapping/core/*
|_ /mapping/internal-txs-processor/*
|_ /mapping/logs-processor/*
|_ /mapping/txs-processor/*
```

File: packages/client/parseServer/src/streams/processor.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  streamConfig: streamConfig[];
+}
+
+const collectionNameBuilder = new CollectionNameBuilder();
+const logsProcessor = new LogsProcessor(collectionNameBuilder);
+const txsProcessor = new TxsProcessor(collectionNameBuilder);
+const internalTxProcessor = new InternalTxsProcessor(collectionNameBuilder);
+
+const verifySignature = (req: Request) => {
+  const ProvidedSignature = req.headers['x-signature'];
```
Comment (b4rtaz) [inline packages/client/parseServer/src/streams/processor.ts:27]: `ProvidedSignature` -> `providedSignature`

File: packages/client/parseServer/src/streams/processor.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      ]),
+    );
+
+    txsMap.set(sync.tag, sync);
+    logsMap.set(sync.tag, sync);
+  }
+  await Promise.all(results);
+};
+
+const InitializeSyncsPlugin = async (parseObject: any, express: Express, apiKey: string, syncs: streamConfig[]) => {
```
Comment (b4rtaz) [inline packages/client/parseServer/src/streams/processor.ts:62]: `InitializeSyncsPlugin` -> `initializeSyncsPlugin`

File: packages/client/parseServer/src/streams/processor.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        }
+      }
+      await Promise.all(results);
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.log('error while inserting logs', e.message);
+      return res.status(500).json({ message: 'error while inserting logs' });
+    }
+
+    return res.status(200).json({ message: 'ok' });
```
Comment (b4rtaz) [inline packages/client/parseServer/src/streams/processor.ts:124]: You may extract this logic to a dedicated file.

**Review by b4rtaz — CHANGES_REQUESTED** [review]
Basically it's OK. I've found a few little things.

File: packages/client/parseServer/src/streams/webbhook.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { IWebhook } from '@moralisweb3/streams-typings';
+import { Express, Request } from 'express';
+import { LogsProcessor, TxsProcessor, CollectionNameBuilder, InternalTxsProcessor, Update } from '@moralisweb3/streams';
+import bodyParser from 'body-parser';
+import Moralis from 'moralis/.';
+import { Upsert } from './upsert';
+
+export const map = new Map();
```
Comment (b4rtaz) [inline packages/client/parseServer/src/streams/webbhook.ts:9]: The name of this variable doesn't say anything about what is this.

**Review by b4rtaz — CHANGES_REQUESTED** [review]

File: packages/parseServer/src/streams/processor.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import { Core } from 'moralis/common-core';
+import { MoralisStreams } from '@moralisweb3/streams';
+import { ApiUtils } from '@moralisweb3/api-utils';
+
+interface StreamOptions {
+  webhookUrl?: string;
+  apiKey: string;
+}
+
+export const streamsSync = (parseInstance: any, options: StreamOptions) => {
```
Comment (b4rtaz) [inline packages/parseServer/src/streams/processor.ts:12]: `parseInstance: any` Can we set here a dedicated type or the parse server doesn't offer TS definitions?

File: packages/parseServer/src/streams/processor.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  const core = Core.create();
+  const streams = MoralisStreams.create(core);
+  const apiUtils = ApiUtils.create(core);
+  core.registerModules([streams, apiUtils]);
+  core.start({ apiKey: options.apiKey });
+  return webhookRouter(parseInstance, options?.webhookUrl || '/streams', streams);
```
Comment (b4rtaz) [inline packages/parseServer/src/streams/processor.ts:18]: I would replace the default path to `/webhook` or `/streams-webhook`.

Comment (b4rtaz) [reply packages/parseServer/src/streams/processor.ts:18]: I think the keyword `webhook` is quite important.

File: packages/parseServer/src/streams/upsert.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import { Document, Update } from '@moralisweb3/streams';
+import Parse from 'parse/node';
+
+export class Upsert {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  constructor(private parseObject: any) {
+    Parse.initialize(this.parseObject.config.appId);
```
Comment (b4rtaz) [inline packages/parseServer/src/streams/upsert.ts:7]: Should we initialize the parse server here again? A user could do it in his own code.

File: packages/parseServer/src/streams/webbhook.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      // eslint-disable-next-line no-console
+      console.log('error while inserting logs', e.message);
+      return res.status(500).json({ message: 'error while inserting logs' });
+    }
+
+    return res.status(200).json({ message: 'ok' });
+  });
+};
+
+const prepareUpdate = (updates: Update[], filters: string[], upsertValue: boolean) => {
```
Comment (b4rtaz) [inline packages/parseServer/src/streams/webbhook.ts:85]: The `upsertValue` argument is unnecessary. The code above passes always `true`.

**Review by b4rtaz — CHANGES_REQUESTED** [review]

File: packages/parseServer/src/streams/upsert.ts (hunk 7 above)
Comment (sogunshola) [reply packages/parseServer/src/streams/upsert.ts:7]: This is the SDK instance.. I'll change `parseObject` to `parseServer` for easier reading

**Pull request closed without merging**